### PR TITLE
Fee query RPC (for fiat_filter calls)

### PR DESCRIPTION
**Pull request by nmrshll**

### CODE REVIEW

**Comment by nmrshll** [comment]
Needs fiat_filter reworked and merged before changing this one from draft to PR

File: pallets/core_mods/src/attest.rs [hunk 1]
```diff
@@ -27,6 +27,12 @@ pub struct Attestation {
     priority: u64,
     iri: Option<Iri>,
 }
+impl Attestation {
```
Comment (lovesh) [inline pallets/core_mods/src/attest.rs:30]: Why is this `impl` added in this PR?

File: pallets/fiat_filter/rpc/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+jsonrpc-core-client = "15.0.0"
+jsonrpc-derive = "15.0.0"
+frame-support = { version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier", default-features = false }
+sp-api = { version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
+sp-runtime = { version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier"}
+sp-blockchain = { version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
+sp-core = { version = "2.0.1", git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
+sp-rpc = { version = "2.0.1", git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
+
+# fiat_filter = { version = '0.1.0', path = '..' }
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/Cargo.toml:19]: Do you intend to keep it?

File: pallets/fiat_filter/rpc/runtime_api/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+name = "fiat_filter_rpc_runtime_api"
+version = "0.0.1"
+authors = ["Dock.io"]
+edition = "2018"
+
+
+[dependencies]
+sp-api = { default-features = false, version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
+sp-runtime = { default-features = false, version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
+# frame-support = { default-features = false, version = '2.0.1', git = "https://github.com/paritytech/substrate.git", branch = "frontier" }
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/runtime_api/Cargo.toml:11]: This commented dependency should be removed.

File: pallets/fiat_filter/rpc/runtime_api/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+codec = { package = 'parity-scale-codec', version = '1.3.6', default-features=false, features=['derive'] }
+
+[features]
+default = ['std']
+std = [
+    'codec/std',
+    # 'frame-support/std',
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/runtime_api/Cargo.toml:18]: Remove commented dependency

File: pallets/fiat_filter/rpc/runtime_api/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+sp_api::decl_runtime_apis! {
+    pub trait FiatFeeRuntimeApi<Balance> where
+        Balance: Codec + MaybeDisplay,
+    {
+        /// Accepts any extrinsic, signature not needed.
+        /// Returns the call fee fee in µDOCK as Balance (Balance is u64 in current runtime)
+        /// If the extrinsic is not one of the calls supported by FiatFilter, returns an error
+        fn get_call_fee_dock(uxt: Block::Extrinsic) -> Result<Balance,Error>;
+        // Block refers to the type argument auto-added to FiatFeeRuntimeApi by the macro
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/runtime_api/src/lib.rs:26]: Comment should go above declaration?

File: pallets/fiat_filter/rpc/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+use std::convert::TryInto;
+use std::sync::Arc;
+
+pub use fiat_filter_rpc_runtime_api::FiatFeeRuntimeApi;
+
+#[rpc]
+pub trait FiatFeeApi<Balance> {
+    /// Accepts a scale-encoded extrinsic, returns fee in µDOCK as Balance (u64)
+    #[rpc(name = "fiat_filter_getCallFeeDock")]
+    fn get_call_fee_dock(&self, encoded_xt: Bytes) -> Result<Balance>;
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/src/lib.rs:20]: RPC's allow you to query state at a specific block (current or past) so that you can know what the fee was at a certain time in past. All (ours and Substrate's that I have seen) current RPCs follow this convention and consumers of this RPC might expect the same. Any reason you wouldn't want to allow this?

File: pallets/fiat_filter/rpc/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            code: ErrorCode::ServerError(FiatFeeRpcError::DecodeError.into()),
+            message: "Failed to decode request".into(),
+            data: Some(format!("{:?}", e).into()),
+        })?;
+
+        // call runtime api method get_call_fee_dock()
+        match api.get_call_fee_dock(&at, uxt) {
+            Ok(rlt) => rlt.map_err(|e| RpcError {
+                code: ErrorCode::ServerError(FiatFeeRpcError::GetCallFeeDock.into()),
+                message: "Failed calling get_call_fee_dock".into(),
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/src/lib.rs:78]: A descriptive message like `Could not get fee in fiat` will be better.

File: pallets/fiat_filter/rpc/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                data: Some(format!("{:?}", e).into()),
+            }),
+            Err(e) => Err(RpcError {
+                code: ErrorCode::ServerError(FiatFeeRpcError::RuntimeError.into()),
+                message: "Failed calling get_call_fee_dock".into(),
```
Comment (lovesh) [inline pallets/fiat_filter/rpc/src/lib.rs:83]: Same as above

File: scripts/rpc/fiat-filter [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# param extrinsic (0x9101040...) is a call to did::new()
```
Comment (lovesh) [inline scripts/rpc/fiat-filter:1]: Is this file needed?

**Review by lovesh — CHANGES_REQUESTED** [review]

File: pallets/fiat_filter/rpc/src/lib.rs (hunk 6 above)
Comment (nmrshll) [reply pallets/fiat_filter/rpc/src/lib.rs:20]: Is there a use case for checking what the fee was in the past ?
I didn't see one when writing the RPC, and thought the only use case would be to check the fee before making a call.
I might be wrong, I can add this parameter if needed.

If you have executed a call, you could see in the chain events how much was actually charged for it, right ?

File: scripts/rpc/fiat-filter (hunk 9 above)
Comment (nmrshll) [reply scripts/rpc/fiat-filter:1]: Not strictly.
Only helpful for manual testing.

File: pallets/fiat_filter/rpc/src/lib.rs (hunk 6 above)
Comment (lovesh) [reply pallets/fiat_filter/rpc/src/lib.rs:17]: Not yet. I just wanted to be consistent with the other RPC calls.

**Review by lovesh — APPROVED** [review]
